### PR TITLE
feat(planner): Add EquiJoinPredicateRule

### LINKED ISSUE #4625: Create planner rule that can analyze a predicate function
The new join package will expose a function that allows users to define a custom predicate function. Having a planner rule that can determine whether or not that predicate function follows a certain pattern would help us leverage some performance gains and avoid undefined behavior.

Specifically, we want to know whether the predicate function defines an equi-join. In other words: is the function simply comparing the equality of columns in the two input tables?

If it is, the planner rule can output a list of the column names being compared in each table, and pass it on to the transformation. This will save the transformation from having to compile the predicate function, which can be expensive.

For the initial implementation, if the predicate function is doing anything besides an equality check between two columns, that should be considered undefined behavior. In this case, the planner rule should somehow signal that to the transformation, which should then return an error.

If it turns out that adding support for non-equi-join predicate functions is easier than we thought, we can file an issue to do it later.

DOD:
- A planner rule exists that can analyze a predicate function to determine whether or not a predicate function contains any operation besides an equality comparison, and respond accordingly.

### PR DESCRIPTION
**Pull request by scbrickley**
This patch adds a few things:

1. A new planner rule called `EquiJoinPredicateRule` that checks if a `join` predicate defines an equi-join (Closes #4625)
2. Stubs for a new `join` package, which will be fully implemented & documented later (See https://github.com/influxdata/flux/issues/4629 and https://github.com/influxdata/flux/issues/4631)
3. A new check in `runtime.Eval` that panics if builtins are not finalized, with a clear error message describing how to fix the issue. This is the result of the difficulty I had trying to evaluate flux scripts in go unit tests within `stdlib`. The solution was to import `fluxinit/static`

I have one concern with the new planner rule that I could use some feedback on: Originally the intent was to have the `join` transformation check what kind of procedure spec it was passed, and to throw an error if it was not an `EquiJoinProcedureSpec`, but due to the scope of #4625, it was easier to just have the planner rule itself return an error. This may increase the workload for adding other implementations of join later on, but not by much. Still, I wonder if it's better to adjust the rule to not throw an error if it doesn't find an equi-join, and instead just return the original procedure spec.

### CODE REVIEW

File: stdlib/join/equijoin_test.go [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+			name: "multiple comparison same column",
+			flux: `import "join"
+			left = from(bucket: "b1", host: "http://localhost:8086")
+				|> filter(fn: (r) => r._measurement == "a")
+			right = from(bucket: "b2", host: "http://localhost:8086")
+				|> filter(fn: (r) => r._measurement == "b")
+			join.join(
+				left: left,
+				right: right,
+				on: (l, r) => r.a == l.b and r.d == l.b,
```
Comment (OfTheDelmer) [inline stdlib/join/equijoin_test.go:187]: Should we explicitly test multiple comparison with an operator error for something like `on: (l, r,) => r.a == l.b and !(r.d == l.b)`?

File: stdlib/join/join.flux [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+// join
+//
+// ## Parameters
+// - left:
+// - right:
+// - on:
+// - as:
+// - method:
+builtin join : (
```
Comment (OfTheDelmer) [inline stdlib/join/join.flux:13]: 🎸

File: stdlib/join/equijoin.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+			}
+			rob, err := getObjectName(rhs)
+			if err != nil {
+				walkErr = err
+				return
+			}
+
+			// Each side of the binary expression should reference either the `l` or `r` object,
+			// but they should not reference the same object.
+			if !((lob == "l") != (rob == "l") && (lob == "r") != (rob == "r")) {
```
[the remark this answers is not in the document]
Comment (scbrickley) [reply stdlib/join/equijoin.go:127]: I don't think so. In plain english, the expression in the code reads as `one side is l and the other side is r`

Yours (I think) reads `both sides are not l OR both sides are not r`

[the remark this answers is not in the document]
Comment (scbrickley) [reply stdlib/join/equijoin.go:127]: Yes, I think so.

File: interpreter/interpreter_test.go [hunk 4]
```diff
@@ -902,3 +902,32 @@ func TestStack(t *testing.T) {
...
+				},
+			},
+		},
+	}
+
+	scope := values.NewScope()
+	scope.Set("r", values.Null)
+
+	intrp := interpreter.NewInterpreter(nil, nil)
+	sideEffects, _ := intrp.Eval(context.Background(), &node, scope, nil)
```
Comment (wolffcm) [inline interpreter/interpreter_test.go:926]: This is great, and it also doesn't quite capture all of the places where we handle `exists`.

We actually evaluate expressions in two places, in the interpreter as you're doing here, and also in the `Compiler` which is a really confusing name. The `Compiler` is the thing that evaluates, e.g., the lambda that's passed into `map`. You could use an acceptance test to do this, if it doesn't produce an error.

Moreover, there's really two kinds of things that should get tested here:
- When we know the names and types of all the fields statically, e.g., what the input data to a function like `map` is produced with `array.from`
- When we don't know all the fields statically, e.g., when the input data is generated via `csv.from`

If I understand write #4626 would need to test both of these cases for both the interpreter and the compiler.

File: runtime/runtime.go [hunk 5]
```diff
@@ -97,6 +97,11 @@ func (r *runtime) Prelude() values.Scope {
 }
 
 func (r *runtime) Eval(ctx context.Context, astPkg flux.ASTHandle, es interpreter.ExecOptsConfig, opts ...flux.ScopeMutator) ([]interpreter.SideEffect, values.Scope, error) {
+	// Give a helpful error message for developers who are trying
+	// to evaluate flux scripts in go unit tests
+	if !r.finalized {
+		panic("runtime is not finalized - consider importing package fluxinit or fluxinit/static")
+	}
```
Comment (wolffcm) [inline runtime/runtime.go:104]: 🎉

File: stdlib/join/equijoin.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+	s := n.ProcedureSpec()
+	spec, ok := s.(*JoinProcedureSpec)
+	if !ok {
+		return nil, false, errors.New(codes.Internal, "invalid spec type on join node")
+	}
+
+	fnBody := spec.On.Fn.Block.Body
+
+	if len(fnBody) != 1 {
+		return nil, false, errors.New(codes.Invalid, "predicate should only contain a single return statement")
```
Comment (wolffcm) [inline stdlib/join/equijoin.go:81]: Users might be confused by this error, since a valid lambda to get passed in might look something like `(l, r) => l.id == r.id and l.foo == r.bar` and doesn't need an explicit `return` statement. (I believe that the semantic graph will have a `Return` statement for that case, but it can be elided in the syntax)

Also, the message should note that there are some forms of lambda that are just not supported yet, so the user understands that they didn't do something wrong, it's just that we have more work to do.

**Review by wolffcm — CHANGES_REQUESTED** [review]
This looks great.

My only issue is that the error messages aren't what I would want users seeing if they began using the new `join` package in anger. These can be hard to get right, especially when there are lots of cases that we can't yet support. We can work on that in this PR, or you could also file an issue to clean it up later.

File: stdlib/join/equijoin.go (hunk 6 above)
Comment (scbrickley) [reply stdlib/join/equijoin.go:81]: Ah, I see your point. I'm open to any suggestions you have for the wording. Maybe something like this?

```
some lambda expressions are not yet supported in join predicates - function body should only contain a single logical expression comparing equality of columns from the left and right tables
```

That seems kinda long winded, but I'm having trouble thinking of a more succinct way to communicate the problem.

File: stdlib/join/equijoin_test.go (hunk 1 above)
Comment (scbrickley) [reply stdlib/join/equijoin_test.go:187]: Yeah, that's probably a good idea. Adding a testcase for that now.

File: stdlib/join/equijoin.go (hunk 6 above)
Comment (scbrickley) [reply stdlib/join/equijoin.go:81]: It would be easier if we could link them to some documentation explaining what is and isn't allowed in a predicate.

Comment (scbrickley) [reply stdlib/join/equijoin.go:81]: Maybe once #4631 is finished, we can circle back and update these errors to just link to the generated fluxdoc page for the new join package.

Comment (wolffcm) [reply stdlib/join/equijoin.go:81]: Sure, I like your message, you could also explicitly refer to the name of the parameter:
```
some expressions are not yet supported in the "on" parameter - function body should only contain a single logical expression comparing equality of columns from the left and right tables
```
(I guess the error message will have prefixed with something like `error in function join` or some such)

**Review by wolffcm — APPROVED** [review]
This looks good, but I don't think that this PR should close #4626, it should stay open until there are more tests.

**Comment by scbrickley** [comment]
Right. Completely forgot about that. I'll move that test to a separate branch and link the issue in a new PR.

**Comment by scbrickley** [comment]
@noramullen1 What are your thoughts on this PR from a docs perspective? There's a new package, but no real docs. Just the bare minimum to get the fluxdoc CI check to pass. There's an open issue to document this package that we intend to get to later: #4631 

I am operating under the assumption that we don't want to document things that are not ready for users yet, but I'm not sure how you and your team feel about this. What should we do here?

**Comment by sanderson** [comment]
@scbrickley I think it's fine to not include full docs at this point. Just make sure that the existing, bare bones docs include something like: "This package is in actively development and is not currently ready to be used". Same for the functions inside the package.

Just in case someone pulls the docs assets from a Flux version where the `join` package isn't ready yet.

**Comment by scbrickley** [comment]
Thanks @sanderson! Making the changes now.
